### PR TITLE
[CAFV-358] Add ZoneDetails object to VCDCluster for Multi-AZ enablement

### PR DESCRIPTION
**Pull request by abaruni**
## Description
Please provide a brief description of the changes proposed in this Pull Request

- 

## Checklist
- [ ] tested locally
- [ ] updated any relevant dependencies
- [ ] updated any relevant documentation or examples

## API Changes
Are there API changes?
- [ ] Yes
- [ ] No

If yes, please fill in the below

1. Updated conversions?
    - [ ] Yes
    - [ ] No
    - [ ] N/A
2. Updated CRDs?
    - [ ] Yes
    - [ ] No
    - [ ] N/A
3. Updated infrastructure-components.yaml?
    - [ ] Yes
    - [ ] No
    - [ ] N/A
4. Updated `./examples/capi-quickstart.yaml`?
    - [ ] Yes
    - [ ] No
    - [ ] N/A
5. Updated necessary files under `./infrastructure-vcd/v1.0.0/`?
   - [ ] Yes
   - [ ] No
   - [ ] N/A

## Issue
If applicable, please reference the relevant issue

Fixes #

- - -
This change is [Reviewable](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/574)

### CODE REVIEW

**Review by sahithi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/574)* status: 0 of 8 files reviewed, 6 unresolved discussions (waiting on @abaruni)

___
*[`api/v1beta3/vcdcluster_types.go` line 93 at r1](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/574#-Nnfhox1A-HwW8NCiiOz:-Nnfhox1A-HwW8NCiiP-:bdt25lv) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/a38bfaad12a7e8e3ea5ec30483b25c18f96c307f/api/v1beta3/vcdcluster_types.go#L93)):*
> ```Go
> 
> 	// +optional
> 	LoadBalancerIP string `json:"loadBalancerIP,omitempty"`
> ```

It is not very evident as to which property needs to be filled in for which zone type and when. Can we combine LoadBalancerIP, LoadBalancerPort, UserSpecifiedEdge properties into a struct called ZoneTopologyDetails? And default values for the struct can be all empty, given default zone type will be "dcgroup".

It will be clear for users that ZoneTopologyDetails will have to be filled in only for the ZoneTopologyTypes - external (or) userspecifiededge.

@arunmk what are your thoughts?
___
*[`api/v1beta3/vcdcluster_types.go` line 101 at r1](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/574#-NnfgwFL-v7deKquWbBk:-NnfgwFM9tvUwPIxliOS:b-56azw3) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/a38bfaad12a7e8e3ea5ec30483b25c18f96c307f/api/v1beta3/vcdcluster_types.go#L101)):*
> ```Go
> type ZoneDetailsSpec struct {
> 	// +optional
> 	ZoneType common.ZoneType `json:"zoneType,omitempty"`
> ```

At first I thought  - Shouldn't _zoneType_ be part of _Zone_ definition by plain english? Reviewed the design doc to refresh my memory on what it actually means.

Wondering if can rename to something better. 
ZoneTopologyType?

@arunmk  what do you think?

can we also add a comment to describe what it means?
___
*[`api/v1beta3/vcdcluster_types.go` line 103 at r1](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/574#-Nnfiraq2iG3f3UwNMjT:-Nnfiraq2iG3f3UwNMjU:b9noyuk) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/a38bfaad12a7e8e3ea5ec30483b25c18f96c307f/api/v1beta3/vcdcluster_types.go#L103)):*
> ```Go
> 	ZoneType common.ZoneType `json:"zoneType,omitempty"`
> 	// +optional
> 	UserSpecifiedEdgeGatewayZone string `json:"userSpecifiedEdgeGatewayZone"`
> ```

A comment here will also be helpful - that this is applicable only when ZoneType is UserSpecifiedEdge.
___
*[`api/v1beta3/vcdcluster_types.go` line 113 at r1](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/574#-NnfnA2pDGqQWvXF9Xk6:-NnfnA2pDGqQWvXF9Xk7:b-8otgnu) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/a38bfaad12a7e8e3ea5ec30483b25c18f96c307f/api/v1beta3/vcdcluster_types.go#L113)):*
> ```Go
> 	ZoneType common.ZoneType `json:"zoneType,omitempty"`
> 	// +optional
> 	UserSpecifiedEdgeGateway string `json:"userSpecifiedEdgeGateway"`
> ```

What is the purpose of this property "UserSpecifiedEdgeGateway". We already have a property called "UserSpecifiedEdgeGatewayZone"?
___
*[`api/v1beta3/vcdcluster_types.go` line 151 at r1](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/574#-NnfqQ9_8wy_SBBTSkxW:-NnfqQ9_8wy_SBBTSkxX:b-fcjbq9) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/a38bfaad12a7e8e3ea5ec30483b25c18f96c307f/api/v1beta3/vcdcluster_types.go#L151)):*
> ```Go
> 	ZoneDetails ZoneDetailsSpec `json:"zoneDetails,omitempty"`
> 	// +optional
> 	ZonesConfigMapName string `json:"zonesConfigMapName,omitempty"`
> ```

Can you add a comment here as to why we need this?
___
*[`api/v1beta3/vcdcluster_types.go` line 206 at r1](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/574#-NnfqGYQ6IWu2-p7Lr3U:-NnfqGYQ6IWu2-p7Lr3V:b-pcjvid) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/a38bfaad12a7e8e3ea5ec30483b25c18f96c307f/api/v1beta3/vcdcluster_types.go#L206)):*
> ```Go
> 
> 	// +optional
> 	FailureDomains clusterv1.FailureDomains `json:"failureDomains,omitempty"`
> ```

For my benefit, who will consume this property?

File: api/v1beta3/vcdcluster_types.go [hunk 1]
```diff
@@ -80,39 +79,112 @@ type LoadBalancerConfig struct {
...
 
+// MultiZoneSpec provides details of the zone configuration in the cluster as well as the NetworkTopology used. Only one
+// of DCGroupConfig, UserSpecifiedEdgeGatewayConfig, or ExternalLoadBalancerConfig should be provided corresponding to the
+// chosen ZoneTopology.
+type MultiZoneSpec struct {
+	// ZoneTopology defines the type of network topology used across zones in a Multi-AZ deployment.
+	// Valid options are DCGroup, UserSpecifiedEdgeGateway, and ExternalLoadBalancer,
 	// +optional
-	LoadBalancerIP string `json:"loadBalancerIP,omitempty"`
+	ZoneTopology ZoneTopologyType `json:"ZoneTopology,omitempty"`
```
Comment (arunmk) [inline api/v1beta3/vcdcluster_types.go:157]: the json representation should start with lowercase (`zoneTopologyType`) in accordance with k8s conventions.

File: api/v1beta3/vcdcluster_types.go [hunk 2]
```diff
@@ -80,39 +79,112 @@ type LoadBalancerConfig struct {
...
-	LoadBalancerPort int `json:"loadBalancerPort,omitempty"`
-}
-
-// ZoneDetailsSpec is a detailed summary of the zones in the cluster as well as the ZoneType used
-type ZoneDetailsSpec struct {
+	DCGroupConfig DCGroupConfig `json:"dcGroupConfig,omitempty"`
+	// UserSpecifiedEdgeGatewayConfig contains configuration for UserSpecifiedEdgeGateway zone topology.
 	// +optional
-	ZoneType common.ZoneType `json:"zoneType,omitempty"`
+	UserSpecifiedEdgeGatewayConfig UserSpecifiedEdgeGatewayConfig `json:"UserSpecifiedEdgeGatewayConfig,omitempty"`
```
Comment (arunmk) [inline api/v1beta3/vcdcluster_types.go:163]: json should start with lowercase

**Review by arunmk — APPROVED** [review]
Thanks @abaruni, this looks good to me.

One small but important comment is that all k8s names (json part of the struct) should use camelBack notation.

**Review by abaruni — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/574)* status: 0 of 11 files reviewed, 9 unresolved discussions (waiting on @arunmk and @sahithi)

___
*[`api/v1beta3/vcdcluster_types.go` line 157 at r2](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/574#gh-1463734461:-NorXNW88zscn_ixqSSk:b-s5nvfh) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/da01f46bb8058858ca57336193830a0d5b4832ca/api/v1beta3/vcdcluster_types.go#L157)):*
<details><summary><i>Previously, arunmk (Arun M. Krishnakumar) wrote…</i></summary><blockquote>

the json representation should start with lowercase (`zoneTopologyType`) in accordance with k8s conventions.
</blockquote></details>

thanks for catching that. just an autosuggestion error
___
*[`api/v1beta3/vcdcluster_types.go` line 163 at r2](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/574#gh-1463735197:-NorXURe6HvmNONWvAEm:b-s5nvfh) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/da01f46bb8058858ca57336193830a0d5b4832ca/api/v1beta3/vcdcluster_types.go#L163)):*
<details><summary><i>Previously, arunmk (Arun M. Krishnakumar) wrote…</i></summary><blockquote>

json should start with lowercase 
</blockquote></details>

thanks for catching that. just an autosuggestion error

**Review by abaruni — COMMENTED** [review]
yes, was just an oversight while accepting code completion suggestion

> *[Reviewable](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/574#-:-NorYEVv3UhCdQ8eTZCZ:bsg9dl9)* status: 0 of 11 files reviewed, 9 unresolved discussions (waiting on @arunmk and @sahithi)

**Comment by Anirudh9794** [comment]
*[`api/v1beta3/vcdcluster_types.go` line 183 at r3](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/574#-Nowf3ilAxtIAqn_uq-j:-Nowf3ilAxtIAqn_uq-k:bivd4at) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/26c42c0a03a9980263015a1ff05a68701d7e5c5c/api/v1beta3/vcdcluster_types.go#L183)):*
> ```Go
> 	// UserSpecifiedEdgeGatewayConfig contains configuration for UserSpecifiedEdgeGateway zone topology.
> 	// +optional
> 	UserSpecifiedEdgeGatewayConfig UserSpecifiedEdgeGatewayConfig `json:"userSpecifiedEdgeGatewayConfig,omitempty"`
> ```

I don't see the fields "dcGroupConfig", "userSpecifiedEdgeGatewayConfig"  or "externalLoadBalancerConfig" fields in `crd/bases`

**Comment by Anirudh9794** [comment]
*[`api/v1beta3/vcdcluster_types.go` line 282 at r3](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/574#-NowbND-AUO46nNmrei1:-NowbND-AUO46nNmrei2:b-276rm5) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/26c42c0a03a9980263015a1ff05a68701d7e5c5c/api/v1beta3/vcdcluster_types.go#L282)):*
> ```Go
> 
> 	// FailureDomains lists the zones of this cluster. This field is parsed from the Zones field of
> 	// vcdCluster.MultiZoneSpec if set up appropriately.
> ```

nit: vcdCluster.MultiZoneSpec -> vcdCluster.Spec.MultiZoneZpec

**Review by Anirudh9794 — APPROVED** [review]

**Review by abaruni — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/574)* status: 0 of 11 files reviewed, 10 unresolved discussions (waiting on @Anirudh9794, @arunmk, and @sahithi)

___
*[`api/v1alpha4/zz_generated.conversion.go` line 341 at r1](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/574#-Nnj8cbQ4aTUZHUfrxrY:-Np5tXTSC1jyxmFWjl2G:b-896fix) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/a38bfaad12a7e8e3ea5ec30483b25c18f96c307f/api/v1alpha4/zz_generated.conversion.go#L341)):*
<details><summary><i>Previously, Anirudh9794 wrote…</i></summary><blockquote>

I think these fields need to be restored in vcdcluster_conversion.go for v1alpha4, v1beta1 and v1beta2
</blockquote></details>

Done.
___
*[`api/v1beta3/vcdcluster_types.go` line 183 at r3](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/574#-Nowf3ilAxtIAqn_uq-j:-Np5tba31d2mzELx-w1d:b-896fix) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/26c42c0a03a9980263015a1ff05a68701d7e5c5c/api/v1beta3/vcdcluster_types.go#L183)):*
<details><summary><i>Previously, Anirudh9794 wrote…</i></summary><blockquote>

I don't see the fields "dcGroupConfig", "userSpecifiedEdgeGatewayConfig"  or "externalLoadBalancerConfig" fields in `crd/bases`
</blockquote></details>

Done.

**Review by abaruni — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/574)* status: 0 of 11 files reviewed, 10 unresolved discussions (waiting on @Anirudh9794, @arunmk, and @sahithi)

___
*[`api/v1beta3/vcdcluster_types.go` line 93 at r1](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/574#-Nnfhox1A-HwW8NCiiOz:-Np5toDB33j0XTug63Mz:b-896fix) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/a38bfaad12a7e8e3ea5ec30483b25c18f96c307f/api/v1beta3/vcdcluster_types.go#L93)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

It is not very evident as to which property needs to be filled in for which zone type and when. Can we combine LoadBalancerIP, LoadBalancerPort, UserSpecifiedEdge properties into a struct called ZoneTopologyDetails? And default values for the struct can be all empty, given default zone type will be "dcgroup".

It will be clear for users that ZoneTopologyDetails will have to be filled in only for the ZoneTopologyTypes - external (or) userspecifiededge.

@arunmk what are your thoughts?
</blockquote></details>

Done.
___
*[`api/v1beta3/vcdcluster_types.go` line 101 at r1](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/574#-NnfgwFL-v7deKquWbBk:-Np5tpeo1aST-Rwq9_Yb:b-896fix) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/a38bfaad12a7e8e3ea5ec30483b25c18f96c307f/api/v1beta3/vcdcluster_types.go#L101)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

At first I thought  - Shouldn't _zoneType_ be part of _Zone_ definition by plain english? Reviewed the design doc to refresh my memory on what it actually means.

Wondering if can rename to something better. 
ZoneTopologyType?

@arunmk  what do you think?

can we also add a comment to describe what it means?
</blockquote></details>

Done.
___
*[`api/v1beta3/vcdcluster_types.go` line 103 at r1](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/574#-Nnfiraq2iG3f3UwNMjT:-Np5trFs3BPTB5gjn5UT:b-896fix) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/a38bfaad12a7e8e3ea5ec30483b25c18f96c307f/api/v1beta3/vcdcluster_types.go#L103)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

A comment here will also be helpful - that this is applicable only when ZoneType is UserSpecifiedEdge.
</blockquote></details>

Done.
___
*[`api/v1beta3/vcdcluster_types.go` line 113 at r1](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/574#-NnfnA2pDGqQWvXF9Xk6:-Np5ts7y81-NV6GVfz6w:b-896fix) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/a38bfaad12a7e8e3ea5ec30483b25c18f96c307f/api/v1beta3/vcdcluster_types.go#L113)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

What is the purpose of this property "UserSpecifiedEdgeGateway". We already have a property called "UserSpecifiedEdgeGatewayZone"?
</blockquote></details>

Done.
___
*[`api/v1beta3/vcdcluster_types.go` line 151 at r1](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/574#-NnfqQ9_8wy_SBBTSkxW:-Np5ttKICF1Tm8lTNZgu:b-896fix) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/a38bfaad12a7e8e3ea5ec30483b25c18f96c307f/api/v1beta3/vcdcluster_types.go#L151)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

Can you add a comment here as to why we need this?
</blockquote></details>

Done.
___
*[`api/v1beta3/vcdcluster_types.go` line 206 at r1](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/574#-NnfqGYQ6IWu2-p7Lr3U:-Np5ttwg1H9lDU6U_i4z:b-896fix) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/a38bfaad12a7e8e3ea5ec30483b25c18f96c307f/api/v1beta3/vcdcluster_types.go#L206)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

For my benefit, who will consume this property?
</blockquote></details>

Done.

**Review by sahithi — COMMENTED** [review]
Last few comments, thanks Adnan.
:lgtm: 

> *[Reviewable](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/574#-:-NosPG853msDMUvtG5bA:bqmwg6a)* status: 0 of 11 files reviewed, 10 unresolved discussions (waiting on @abaruni, @Anirudh9794, and @arunmk)

___
*[`api/v1beta3/vcdcluster_types.go` line 151 at r1](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/574#-NnfqQ9_8wy_SBBTSkxW:-NosBm364e4dWroAdbeI:b3ba062) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/a38bfaad12a7e8e3ea5ec30483b25c18f96c307f/api/v1beta3/vcdcluster_types.go#L151)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

Can you add a comment here as to why we need this?
</blockquote></details>

OVDCZoneConfigMap comment can be elaborated as this field dictates how CPI, CSI need to be configured in order for all these components to work in harmony in a MultiAZ setup.

At this point it is not clear why user needs to enter it, how it is going to get populated, where else user needs to use this name?
___
*[`api/v1beta3/vcdcluster_types.go` line 173 at r3](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/574#-NosOCgA4ZNCcSaOGCj-:-NosOCgA4ZNCcSaOGCj0:bl2fzq0) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/26c42c0a03a9980263015a1ff05a68701d7e5c5c/api/v1beta3/vcdcluster_types.go#L173)):*
> ```Go
> 
> // MultiZoneStatus provides the current status of the zone configuration in a Multi-AZ deployment.
> type MultiZoneStatus struct {
> ```

OvdcZoneConfigMap missing here?
___
*[`config/crd/bases/infrastructure.cluster.x-k8s.io_vcdclusters.yaml` line 681 at r1](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/574#-NnfqqJA6GZOP7p3KMQL:-NnfqqJA6GZOP7p3KMQM:b-qa30y0) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/a38bfaad12a7e8e3ea5ec30483b25c18f96c307f/config/crd/bases/infrastructure.cluster.x-k8s.io_vcdclusters.yaml#L681)):*
> ```YAML
>                 type: object
>               zoneDetails:
>                 description: ZoneDetailsSpec is a detailed summary of the zones in
> ```

Curious to know how is the description generated?

**Review by abaruni — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/574)* status: 0 of 11 files reviewed, 10 unresolved discussions (waiting on @Anirudh9794, @arunmk, and @sahithi)

___
*[`config/crd/bases/infrastructure.cluster.x-k8s.io_vcdclusters.yaml` line 681 at r1](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/574#-NnfqqJA6GZOP7p3KMQL:-NpLpIPo8xFr5dXFYltz:blcqzk8) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/a38bfaad12a7e8e3ea5ec30483b25c18f96c307f/config/crd/bases/infrastructure.cluster.x-k8s.io_vcdclusters.yaml#L681)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

Curious to know how is the description generated?
</blockquote></details>

it is generated from the comments on the field

**Pull request merged**